File: src/reffy/storage.py
from __future__ import annotations

import json
from dataclasses import asdict, dataclass
from datetime import datetime, timezone
from pathlib import Path
import mimetypes
from typing import Any
from uuid import uuid4

MANIFEST_VERSION = 1


def _utc_now() -> str:
    return datetime.now(timezone.utc).isoformat()


@dataclass
class Artifact:
    id: str
    name: str
    filename: str
    kind: str
    mime_type: str
    size_bytes: int
    tags: list[str]
    created_at: str
    updated_at: str


class ReferencesStore:
    def __init__(self, repo_root: Path) -> None:
        self.repo_root = repo_root
        self.refs_dir = repo_root / ".references"
        self.artifacts_dir = self.refs_dir / "artifacts"
        self.manifest_path = self.refs_dir / "manifest.json"
        self._ensure_structure()

    def _ensure_structure(self) -> None:
        self.refs_dir.mkdir(parents=True, exist_ok=True)
        self.artifacts_dir.mkdir(parents=True, exist_ok=True)
        if not self.manifest_path.exists():
            self._write_manifest(self._empty_manifest())

    def _empty_manifest(self) -> dict[str, Any]:
        now = _utc_now()
        return {
            "version": MANIFEST_VERSION,
            "created_at": now,
            "updated_at": now,
            "artifacts": [],
            "conflicts": [],
        }

    def _read_manifest(self) -> dict[str, Any]:
        try:
            data = json.loads(self.manifest_path.read_text())
        except FileNotFoundError:
            data = self._empty_manifest()
        if isinstance(data, list):
            # Backward-compat: older manifest stored only a list of artifacts.
            data = {
                "version": 0,
                "created_at": _utc_now(),
                "updated_at": _utc_now(),
                "artifacts": data,
            }
        if not isinstance(data, dict):
            return self._empty_manifest()
        data.setdefault("version", MANIFEST_VERSION)
        data.setdefault("created_at", _utc_now())
        data.setdefault("updated_at", _utc_now())
        data.setdefault("artifacts", [])
        data.setdefault("conflicts", [])
        return data

    def _write_manifest(self, manifest: dict[str, Any]) -> None:
        self.manifest_path.write_text(json.dumps(manifest, indent=2, sort_keys=True))

    def list_artifacts(self) -> list[dict[str, Any]]:
        return self._read_manifest().get("artifacts", [])

    def get_manifest(self) -> dict[str, Any]:
        return self._read_manifest()

    def get_artifact(self, artifact_id: str) -> dict[str, Any] | None:
        for item in self._read_manifest().get("artifacts", []):
            if item.get("id") == artifact_id:
                return item
        return None

    def get_artifact_path(self, artifact: dict[str, Any]) -> Path:
        return self.artifacts_dir / artifact["filename"]

    def _slugify(self, name: str) -> str:
        cleaned = "".join(ch if ch.isalnum() or ch in ("-", "_", " ") else "" for ch in name)
        cleaned = "-".join(part for part in cleaned.strip().split() if part)
        return cleaned.lower() or "untitled"

    def _unique_filename(self, base: str, ext: str = ".md") -> str:
        candidate = f"{base}{ext}"
        if not (self.artifacts_dir / candidate).exists():
            return candidate
        counter = 2
        while True:
            candidate = f"{base}-{counter}{ext}"
            if not (self.artifacts_dir / candidate).exists():
                return candidate
            counter += 1

    def create_artifact(
        self,
        name: str,
        content: str | None,
        *,
        kind: str | None = None,
        mime_type: str | None = None,
        tags: list[str] | None = None,
    ) -> dict[str, Any]:
        artifact_id = str(uuid4())
        safe_name = self._slugify(name)
        filename = self._unique_filename(safe_name, ".md")
        now = _utc_now()
        path = self.artifacts_dir / filename
        if content is not None:
            path.write_text(content)
        size_bytes = path.stat().st_size if path.exists() else 0
        artifact = Artifact(
            id=artifact_id,
            name=name,
            filename=filename,
            kind=kind or "note",
            mime_type=mime_type or "text/markdown",
            size_bytes=size_bytes,
            tags=tags or [],
            created_at=now,
            updated_at=now,
        )
        manifest = self._read_manifest()
        manifest["updated_at"] = _utc_now()
        manifest["artifacts"].append(asdict(artifact))
        self._write_manifest(manifest)
        return asdict(artifact)

    def _infer_kind(self, path: Path) -> tuple[str, str]:
        suffix = path.suffix.lower()
        if suffix == ".excalidraw":
            return "diagram", "application/json"
        if suffix in {".png", ".jpg", ".jpeg"}:
            mime = "image/png" if suffix == ".png" else "image/jpeg"
            return "image", mime
        if suffix in {".html", ".htm"}:
            return "html", "text/html"
        if suffix == ".pdf":
            return "pdf", "application/pdf"
        if suffix == ".doc":
            return "doc", "application/msword"
        if suffix == ".docx":
            return "doc", "application/vnd.openxmlformats-officedocument.wordprocessingml.document"
        if suffix == ".json":
            return "json", "application/json"
        if suffix == ".md":
            return "note", "text/markdown"
        mime = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
        return "file", mime

    def reindex_artifacts(self) -> dict[str, Any]:
        manifest = self._read_manifest()
        artifacts = manifest.get("artifacts", [])
        known_filenames = {item.get("filename") for item in artifacts if item.get("filename")}
        added = 0
        for path in self.artifacts_dir.iterdir():
            if path.is_dir():
                continue
            if path.name in known_filenames:
                continue
            kind, mime_type = self._infer_kind(path)
            now = _utc_now()
            artifact = Artifact(
                id=str(uuid4()),
                name=path.stem.replace("-", " ").strip() or "untitled",
                filename=path.name,
                kind=kind,
                mime_type=mime_type,
                size_bytes=path.stat().st_size,
                tags=[],
                created_at=now,
                updated_at=now,
            )
            artifacts.append(asdict(artifact))
            added += 1
        if added:
            manifest["artifacts"] = artifacts
            manifest["updated_at"] = _utc_now()
            self._write_manifest(manifest)
        return {"added": added, "total": len(artifacts)}

    def update_artifact(
        self,
        artifact_id: str,
        *,
        name: str | None = None,
        content: str | None = None,
        kind: str | None = None,
        mime_type: str | None = None,
        tags: list[str] | None = None,
    ) -> dict[str, Any] | None:
        manifest = self._read_manifest()
        items = manifest.get("artifacts", [])
        for idx, item in enumerate(items):
            if item.get("id") != artifact_id:
                continue
            if name is not None:
                item["name"] = name
            if kind is not None:
                item["kind"] = kind
            if mime_type is not None:
                item["mime_type"] = mime_type
            if tags is not None:
                item["tags"] = tags
            if content is not None:
                path = self.get_artifact_path(item)
                path.write_text(content)
                item["size_bytes"] = path.stat().st_size
            item["updated_at"] = _utc_now()
            items[idx] = item
            manifest["updated_at"] = _utc_now()
            self._write_manifest(manifest)
            return item
        return None

    def delete_artifact(self, artifact_id: str) -> bool:
        manifest = self._read_manifest()
        items = manifest.get("artifacts", [])
        remaining = []
        removed = None
        for item in items:
            if item.get("id") == artifact_id:
                removed = item
            else:
                remaining.append(item)
        if not removed:
            return False
        path = self.get_artifact_path(removed)
        if path.exists():
            path.unlink()
        manifest["artifacts"] = remaining
        manifest["updated_at"] = _utc_now()
        self._write_manifest(manifest)
        return True

    def record_conflict(
        self,
        *,
        artifact_id: str,
        source: str,
        note: str,
        conflict_artifact_id: str | None = None,
    ) -> dict[str, Any]:
        manifest = self._read_manifest()
        entry = {
            "artifact_id": artifact_id,
            "source": source,
            "note": note,
            "conflict_artifact_id": conflict_artifact_id,
            "created_at": _utc_now(),
        }
        manifest["updated_at"] = _utc_now()
        manifest["conflicts"].append(entry)
        self._write_manifest(manifest)
        return entry

    def create_conflict_copy(
        self,
        *,
        artifact_id: str,
        source: str,
        note: str,
    ) -> dict[str, Any] | None:
        original = self.get_artifact(artifact_id)
        if not original:
            return None
        original_path = self.get_artifact_path(original)
        if not original_path.exists():
            return None
        content = original_path.read_text()
        conflict_name = f"{original['name']} (conflict)"
        conflict_artifact = self.create_artifact(
            name=conflict_name,
            content=content,
            kind=original.get("kind") or "note",
            mime_type=original.get("mime_type") or "text/markdown",
            tags=(original.get("tags") or []) + ["conflict"],
        )
        self.record_conflict(
            artifact_id=artifact_id,
            source=source,
            note=note,
            conflict_artifact_id=conflict_artifact["id"],
        )
        return conflict_artifact

File: src/storage.ts
import { randomUUID } from "node:crypto";
import { mkdirSync } from "node:fs";
import { promises as fs } from "node:fs";
import path from "node:path";
import { lookup as mimeLookup } from "mime-types";

import type { Artifact, Manifest } from "./types.js";

const MANIFEST_VERSION = 1;

function utcNow(): string {
  return new Date().toISOString();
}

function isObject(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

export class ReferencesStore {
  public readonly repoRoot: string;
  public readonly refsDir: string;
  public readonly artifactsDir: string;
  public readonly manifestPath: string;

  constructor(repoRoot: string) {
    this.repoRoot = repoRoot;
    this.refsDir = path.join(repoRoot, ".references");
    this.artifactsDir = path.join(this.refsDir, "artifacts");
    this.manifestPath = path.join(this.refsDir, "manifest.json");
    this.ensureStructure();
  }

  private ensureStructure(): void {
    mkdirSync(this.refsDir, { recursive: true });
    mkdirSync(this.artifactsDir, { recursive: true });
    void fs.access(this.manifestPath).catch(async () => {
      await this.writeManifest(this.emptyManifest());
    });
  }

  private emptyManifest(): Manifest {
    const now = utcNow();
    return {
      version: MANIFEST_VERSION,
      created_at: now,
      updated_at: now,
      artifacts: [],
      conflicts: [],
    };
  }

  private async readManifest(): Promise<Manifest> {
    let raw: unknown;
    try {
      const text = await fs.readFile(this.manifestPath, "utf8");
      raw = JSON.parse(text) as unknown;
    } catch {
      return this.emptyManifest();
    }

    if (Array.isArray(raw)) {
      return {
        version: 0,
        created_at: utcNow(),
        updated_at: utcNow(),
        artifacts: raw as Artifact[],
        conflicts: [],
      };
    }

    if (!isObject(raw)) {
      return this.emptyManifest();
    }

    const artifacts = Array.isArray(raw.artifacts) ? (raw.artifacts as Artifact[]) : [];
    const conflicts = Array.isArray(raw.conflicts) ? raw.conflicts : [];

    return {
      version: typeof raw.version === "number" ? raw.version : MANIFEST_VERSION,
      created_at: typeof raw.created_at === "string" ? raw.created_at : utcNow(),
      updated_at: typeof raw.updated_at === "string" ? raw.updated_at : utcNow(),
      artifacts,
      conflicts: conflicts as Manifest["conflicts"],
    };
  }

  private async writeManifest(manifest: Manifest): Promise<void> {
    await fs.writeFile(this.manifestPath, JSON.stringify(manifest, null, 2));
  }

  public async listArtifacts(): Promise<Artifact[]> {
    const manifest = await this.readManifest();
    return manifest.artifacts;
  }

  public async getArtifact(artifactId: string): Promise<Artifact | null> {
    const manifest = await this.readManifest();
    return manifest.artifacts.find((item) => item.id === artifactId) ?? null;
  }

  public getArtifactPath(artifact: Artifact): string {
    return path.join(this.artifactsDir, artifact.filename);
  }

  private slugify(name: string): string {
    const cleaned = name
      .split("")
      .filter((ch) => /[\w\- ]/.test(ch))
      .join("")
      .trim()
      .replace(/\s+/g, "-")
      .toLowerCase();
    return cleaned || "untitled";
  }

  private async uniqueFilename(base: string, ext = ".md"): Promise<string> {
    let candidate = `${base}${ext}`;
    let counter = 2;
    while (true) {
      try {
        await fs.access(path.join(this.artifactsDir, candidate));
        candidate = `${base}-${counter}${ext}`;
        counter += 1;
      } catch {
        return candidate;
      }
    }
  }

  public async createArtifact(input: {
    name: string;
    content?: string | null;
    kind?: string | null;
    mime_type?: string | null;
    tags?: string[] | null;
  }): Promise<Artifact> {
    const id = randomUUID();
    const safe = this.slugify(input.name);
    const filename = await this.uniqueFilename(safe, ".md");
    const now = utcNow();
    const artifactPath = path.join(this.artifactsDir, filename);

    if (input.content !== undefined && input.content !== null) {
      await fs.writeFile(artifactPath, input.content, "utf8");
    }

    let sizeBytes = 0;
    try {
      const stat = await fs.stat(artifactPath);
      sizeBytes = stat.size;
    } catch {
      sizeBytes = 0;
    }

    const artifact: Artifact = {
      id,
      name: input.name,
      filename,
      kind: input.kind ?? "note",
      mime_type: input.mime_type ?? "text/markdown",
      size_bytes: sizeBytes,
      tags: input.tags ?? [],
      created_at: now,
      updated_at: now,
    };

    const manifest = await this.readManifest();
    manifest.updated_at = utcNow();
    manifest.artifacts.push(artifact);
    await this.writeManifest(manifest);
    return artifact;
  }

  private inferKind(filePath: string): { kind: string; mime_type: string } {
    const ext = path.extname(filePath).toLowerCase();
    if (ext === ".excalidraw") return { kind: "diagram", mime_type: "application/json" };
    if (ext === ".png") return { kind: "image", mime_type: "image/png" };
    if (ext === ".jpg" || ext === ".jpeg") return { kind: "image", mime_type: "image/jpeg" };
    if (ext === ".html" || ext === ".htm") return { kind: "html", mime_type: "text/html" };
    if (ext === ".pdf") return { kind: "pdf", mime_type: "application/pdf" };
    if (ext === ".doc") return { kind: "doc", mime_type: "application/msword" };
    if (ext === ".docx") {
      return {
        kind: "doc",
        mime_type: "application/vnd.openxmlformats-officedocument.wordprocessingml.document",
      };
    }
    if (ext === ".json") return { kind: "json", mime_type: "application/json" };
    if (ext === ".md") return { kind: "note", mime_type: "text/markdown" };
    const guessed = mimeLookup(path.basename(filePath)) || "application/octet-stream";
    return { kind: "file", mime_type: guessed.toString() };
  }

  public async reindexArtifacts(): Promise<{ added: number; total: number }> {
    const manifest = await this.readManifest();
    const known = new Set(manifest.artifacts.map((a) => a.filename));
    const entries = await fs.readdir(this.artifactsDir, { withFileTypes: true });
    let added = 0;

    for (const entry of entries) {
      if (!entry.isFile()) continue;
      if (known.has(entry.name)) continue;
      const filePath = path.join(this.artifactsDir, entry.name);
      const stats = await fs.stat(filePath);
      const now = utcNow();
      const inferred = this.inferKind(filePath);
      const artifact: Artifact = {
        id: randomUUID(),
        name: path.basename(entry.name, path.extname(entry.name)).replace(/-/g, " ").trim() || "untitled",
        filename: entry.name,
        kind: inferred.kind,
        mime_type: inferred.mime_type,
        size_bytes: stats.size,
        tags: [],
        created_at: now,
        updated_at: now,
      };
      manifest.artifacts.push(artifact);
      added += 1;
    }

    if (added > 0) {
      manifest.updated_at = utcNow();
      await this.writeManifest(manifest);
    }

    return { added, total: manifest.artifacts.length };
  }

  public async updateArtifact(
    artifactId: string,
    input: {
      name?: string | null;
      content?: string | null;
      kind?: string | null;
      mime_type?: string | null;
      tags?: string[] | null;
    },
  ): Promise<Artifact | null> {
    const manifest = await this.readManifest();
    const index = manifest.artifacts.findIndex((item) => item.id === artifactId);
    if (index === -1) return null;

    const item = manifest.artifacts[index];
    if (input.name !== undefined && input.name !== null) item.name = input.name;
    if (input.kind !== undefined && input.kind !== null) item.kind = input.kind;
    if (input.mime_type !== undefined && input.mime_type !== null) item.mime_type = input.mime_type;
    if (input.tags !== undefined && input.tags !== null) item.tags = input.tags;

    if (input.content !== undefined && input.content !== null) {
      const artifactPath = this.getArtifactPath(item);
      await fs.writeFile(artifactPath, input.content, "utf8");
      const stat = await fs.stat(artifactPath);
      item.size_bytes = stat.size;
    }

    item.updated_at = utcNow();
    manifest.updated_at = utcNow();
    manifest.artifacts[index] = item;
    await this.writeManifest(manifest);
    return item;
  }

  public async deleteArtifact(artifactId: string): Promise<boolean> {
    const manifest = await this.readManifest();
    const index = manifest.artifacts.findIndex((item) => item.id === artifactId);
    if (index === -1) return false;

    const [removed] = manifest.artifacts.splice(index, 1);
    const artifactPath = this.getArtifactPath(removed);
    await fs.rm(artifactPath, { force: true });

    manifest.updated_at = utcNow();
    await this.writeManifest(manifest);
    return true;
  }

  public async recordConflict(input: {
    artifact_id: string;
    source: string;
    note: string;
    conflict_artifact_id?: string | null;
  }): Promise<void> {
    const manifest = await this.readManifest();
    manifest.conflicts.push({
      artifact_id: input.artifact_id,
      source: input.source,
      note: input.note,
      conflict_artifact_id: input.conflict_artifact_id ?? null,
      created_at: utcNow(),
    });
    manifest.updated_at = utcNow();
    await this.writeManifest(manifest);
  }

  public async createConflictCopy(input: {
    artifact_id: string;
    source: string;
    note: string;
  }): Promise<Artifact | null> {
    const original = await this.getArtifact(input.artifact_id);
    if (!original) return null;

    const originalPath = this.getArtifactPath(original);
    let content: string;
    try {
      content = await fs.readFile(originalPath, "utf8");
    } catch {
      return null;
    }

    const conflict = await this.createArtifact({
      name: `${original.name} (conflict)`,
      content,
      kind: original.kind,
      mime_type: original.mime_type,
      tags: [...original.tags, "conflict"],
    });

    await this.recordConflict({
      artifact_id: input.artifact_id,
      source: input.source,
      note: input.note,
      conflict_artifact_id: conflict.id,
    });

    return conflict;
  }
}
